test/logstream/kubectl.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logstream
+
+import (
+	"bufio"
+	"encoding/json"
+	"log"
+	"os/exec"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/knative/serving/test"
+)
+
+type kubectl struct {
+	Namespace string
+
+	m    sync.RWMutex
+	keys map[string]logger
+}
+
+type logger func(string, ...interface{})
+
+var _ streamer = (*kubectl)(nil)
+
+// Init implements streamer
+func (k *kubectl) Init() {
+	// Start to stream logs, parsing each line as JSON and looking keys that match ours.
+	cmdName := "kubectl"
+	cmdArgs := []string{"logs", "-f", "-n" + k.Namespace, "-lapp", "--max-log-requests=100"}
+
+	cmd := exec.Command(cmdName, cmdArgs...)
+	cmdReader, err := cmd.StdoutPipe()
+	if err != nil {
+		log.Fatalf("error creating Cmd stdout pipe: %v", err)
+	}
+	errReader, err := cmd.StderrPipe()
+	if err != nil {
+		log.Fatalf("error creating Cmd stderr pipe: %v", err)
+	}
+
+	go func() {
+		scanner := bufio.NewScanner(cmdReader)
+		for scanner.Scan() {
+			k.handleLine(scanner.Text())
+		}
+	}()
+
+	go func() {
+		scanner := bufio.NewScanner(errReader)
+		for scanner.Scan() {
+			log.Print(scanner.Text())
+		}
+	}()
+
+	err = cmd.Start()
+	if err != nil {
+		log.Fatalf("error starting Cmd: %v", err)
+	}
+
+	err = cmd.Wait()
+	if err != nil {
+		log.Fatalf("error waiting for Cmd: %v", err)
+	}
+}
+
+// This holds the standard structure of our logs.
+type line struct {
+	Level      string    `json:"level"`
+	Timestamp  time.Time `json:"ts"`
+	Controller string    `json:"knative.dev/controller"`
+	Caller     string    `json:"caller"`
+	Key        string    `json:"knative.dev/key"`
+	Message    string    `json:"msg"`
+
+	// TODO(mattmoor): Parse out more context.
+}
+
+func (k *kubectl) handleLine(l string) {
+	var line line
+	if err := json.Unmarshal([]byte(l), &line); err != nil {
+		// log.Printf("error parsing line: %v; %s", err, l)
+		return
```

Comment:
why silently fail here?

Reply:
There are logs that come through from various libraries that don't have the JSON structure.  I think this would be very noisy for cases with mesh, where we inject sidecars into a couple pods (Envoy doesn't use our structured format).